lib/streamlit/elements/lib/dialog.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import time
+from types import TracebackType
+from typing import Literal, cast
+
+from typing_extensions import TypeAlias
+
+from streamlit.cursor import Cursor
+from streamlit.delta_generator import DeltaGenerator, _enqueue_message
+from streamlit.proto.Block_pb2 import Block as BlockProto
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+DialogWidth: TypeAlias = Literal["small", "large"]
+
+
+class Dialog(DeltaGenerator):
+    @staticmethod
+    def _create(
+        parent: DeltaGenerator,
+        title: str,
+        *,
+        dismissible: bool = True,
+        width: DialogWidth = "small",
+    ) -> Dialog:
+        block_proto = BlockProto()
+        block_proto.dialog.title = title
+        block_proto.dialog.dismissible = dismissible
+        block_proto.dialog.width = str(width)
+
+        delta_path: list[int] = (
+            parent._active_dg._cursor.delta_path if parent._active_dg._cursor else []
+        )
+        dialog = cast(Dialog, parent._block(block_proto=block_proto, dg_type=Dialog))
+
+        dialog._delta_path = delta_path
+        dialog._current_proto = block_proto
+        # We add a sleep here to give the web app time to react to the update. Otherwise,
+        #  we might run into issues where the dialog cannot be opened again after closing
+        time.sleep(0.05)
```

Comment:
I'm a bit confused by this? What's the scenario we are avoiding?

Reply:
Yeah I am also not too happy with that, but without the sleep there seems to be an issue where if the dialog is dismissed - which is a UI-only event - and the dialog being opened again via a click, the opening might not happen. I assume the issue is that the proto message is sent too early and interferes with the React updates. The sleeping seemed to have solved the issue. If I remember correctly, `st.expander` is doing sth. similar. Another alternative that might be better could be to make the dialog a bidirectional component and communicate the dismiss to the backend for easier syncing. Do you think that would be the better approach or do you have another good idea?

Reply:
I'm probably okay with this (I didn't realize this is a relatively common pattern. Would love to file tech debt of sorts to organize this better into a better abstraction.